diff -Paur --no-dereference -- hello.upstream/Makefile.in hello/Makefile.in
--- hello.upstream/Makefile.in
+++ hello/Makefile.in
@@ -1773,7 +1773,7 @@
 # Subdirectories to descend into.
 SUBDIRS = po
 hello_SOURCES = src/hello.c src/system.h
-hello_LDADD = $(LIBINTL) $(top_builddir)/lib/lib$(PACKAGE).a
+hello_LDADD = $(top_builddir)/lib/lib$(PACKAGE).a $(LIBINTL)
 CLEANFILES = $(am__append_2) lib/configmake.h lib/configmake.h-t
 MAINTAINERCLEANFILES = lib/iconv_open-aix.h lib/iconv_open-hpux.h \
 	lib/iconv_open-irix.h lib/iconv_open-osf.h \
diff -Paur --no-dereference -- hello.upstream/src/hello.c hello/src/hello.c
--- hello.upstream/src/hello.c
+++ hello/src/hello.c
@@ -163,7 +163,7 @@
     error (EXIT_FAILURE, errno, _("conversion to a multibyte string failed"));
 
   /* Print greeting message and exit. */
-  wprintf (L"%ls\n", mb_greeting);
+  printf ("%ls\n", mb_greeting);
   free(mb_greeting);
 
   exit (EXIT_SUCCESS);
