diff -Paur --no-dereference -- libssl.upstream/apps/nc/netcat.c libssl/apps/nc/netcat.c
--- libssl.upstream/apps/nc/netcat.c
+++ libssl/apps/nc/netcat.c
@@ -40,7 +40,10 @@
 #include <netinet/in.h>
 #include <netinet/tcp.h>
 #include <netinet/ip.h>
+/* PATCH: Sortix doesn't have <netinet/telnet.h> yet. */
+#if __has_include(<netinet/telnet.h>)
 #include <arpa/telnet.h>
+#endif
 
 #include <ctype.h>
 #include <err.h>
@@ -467,8 +470,12 @@
 		} else {
 			strlcpy(unix_dg_tmp_socket_buf, "/tmp/nc.XXXXXXXXXX",
 			    UNIX_DG_TMP_SOCKET_SIZE);
-			if (mktemp(unix_dg_tmp_socket_buf) == NULL)
+			/* PATCH: Sortix doesn't have the obsolete mktemp(3) function. */
+			int fd = mkstemp(unix_dg_tmp_socket_buf);
+			if (fd < 0)
 				err(1, "mktemp");
+			unlink(unix_dg_tmp_socket_buf);
+			close(fd);
 			unix_dg_tmp_socket = unix_dg_tmp_socket_buf;
 		}
 	}
@@ -1385,7 +1392,8 @@
 	memset(&cmsgbuf, 0, sizeof(cmsgbuf));
 	memset(&iov, 0, sizeof(iov));
 
-	mh.msg_control = (caddr_t)&cmsgbuf.buf;
+	/* PATCH: Sortix doesn't have the non-standard caddr_t type. */
+	mh.msg_control = (char*)&cmsgbuf.buf;
 	mh.msg_controllen = sizeof(cmsgbuf.buf);
 	cmsg = CMSG_FIRSTHDR(&mh);
 	cmsg->cmsg_len = CMSG_LEN(sizeof(int));
@@ -1422,6 +1430,7 @@
 void
 atelnet(int nfd, unsigned char *buf, unsigned int size)
 {
+#ifdef IAC
 	unsigned char *p, *end;
 	unsigned char obuf[4];
 
@@ -1447,6 +1456,9 @@
 		if (atomicio(vwrite, nfd, obuf, 3) != 3)
 			warn("Write Error!");
 	}
+#else
+	errx(1, "Telnet negotation is not supported");
+#endif
 }
 
 int
@@ -1590,16 +1602,20 @@
 			err(1, NULL);
 	}
 	if (Tflag != -1) {
+/* PATCH: The Sortix network stack is WIP and doesn't have IP_TOS yet and
+          likewise with other features. */
+#ifdef IP_TOS
 		if (af == AF_INET && setsockopt(s, IPPROTO_IP,
 		    IP_TOS, &Tflag, sizeof(Tflag)) == -1)
 			err(1, "set IP ToS");
+#endif
 
 #ifdef IPV6_TCLASS
-		else if (af == AF_INET6 && setsockopt(s, IPPROTO_IPV6,
+		if (af == AF_INET6 && setsockopt(s, IPPROTO_IPV6,
 		    IPV6_TCLASS, &Tflag, sizeof(Tflag)) == -1)
 			err(1, "set IPv6 traffic class");
 #else
-		else if (af == AF_INET6) {
+		if (af == AF_INET6) {
 			errno = ENOPROTOOPT;
 			err(1, "set IPv6 traffic class not supported");
 		}
@@ -1617,13 +1633,16 @@
 	}
 
 	if (ttl != -1) {
+#ifdef IP_TTL
 		if (af == AF_INET && setsockopt(s, IPPROTO_IP,
 		    IP_TTL, &ttl, sizeof(ttl)))
 			err(1, "set IP TTL");
-
-		else if (af == AF_INET6 && setsockopt(s, IPPROTO_IPV6,
+#endif
+#ifdef IPV6_UNICAST_HOPS
+		if (af == AF_INET6 && setsockopt(s, IPPROTO_IPV6,
 		    IPV6_UNICAST_HOPS, &ttl, sizeof(ttl)))
 			err(1, "set IPv6 unicast hops");
+#endif
 	}
 
 	if (minttl != -1) {
@@ -1661,7 +1680,9 @@
 		{ "af41",		IPTOS_DSCP_AF41 },
 		{ "af42",		IPTOS_DSCP_AF42 },
 		{ "af43",		IPTOS_DSCP_AF43 },
+#ifdef IPTOS_PREC_CRITIC_ECP
 		{ "critical",		IPTOS_PREC_CRITIC_ECP },
+#endif
 		{ "cs0",		IPTOS_DSCP_CS0 },
 		{ "cs1",		IPTOS_DSCP_CS1 },
 		{ "cs2",		IPTOS_DSCP_CS2 },
@@ -1671,11 +1692,21 @@
 		{ "cs6",		IPTOS_DSCP_CS6 },
 		{ "cs7",		IPTOS_DSCP_CS7 },
 		{ "ef",			IPTOS_DSCP_EF },
+#ifdef IPTOS_PREC_INTERNETCONTROL
 		{ "inetcontrol",	IPTOS_PREC_INTERNETCONTROL },
+#endif
+#ifdef IPTOS_LOWDELAY
 		{ "lowdelay",		IPTOS_LOWDELAY },
+#endif
+#ifdef IPTOS_PREC_NETCONTROL
 		{ "netcontrol",		IPTOS_PREC_NETCONTROL },
+#endif
+#ifdef IPTOS_RELIABILITY
 		{ "reliability",	IPTOS_RELIABILITY },
+#endif
+#ifdef IPTOS_THROUGHPUT
 		{ "throughput",		IPTOS_THROUGHPUT },
+#endif
 		{ NULL,			-1 },
 	};
 
@@ -1761,10 +1792,19 @@
 	if (tls_peer_cert_issuer(tls_ctx))
 		fprintf(stderr, "Issuer: %s\n",
 		    tls_peer_cert_issuer(tls_ctx));
-	if ((t = tls_peer_cert_notbefore(tls_ctx)) != -1)
+	// PATCH: Sortix time functions use TAI-10 instead of UTC.
+	if ((t = tls_peer_cert_notbefore(tls_ctx)) != -1) {
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		add_leap_seconds(&t);
+#endif
 		fprintf(stderr, "Valid From: %s", ctime(&t));
-	if ((t = tls_peer_cert_notafter(tls_ctx)) != -1)
+	}
+	if ((t = tls_peer_cert_notafter(tls_ctx)) != -1) {
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		add_leap_seconds(&t);
+#endif
 		fprintf(stderr, "Valid Until: %s", ctime(&t));
+	}
 	if (tls_peer_cert_hash(tls_ctx))
 		fprintf(stderr, "Cert Hash: %s\n",
 		    tls_peer_cert_hash(tls_ctx));
@@ -1782,12 +1822,21 @@
 		    tls_peer_ocsp_cert_status(tls_ctx),
 		    tls_peer_ocsp_crl_reason(tls_ctx));
 		t = tls_peer_ocsp_this_update(tls_ctx);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		add_leap_seconds(&t);
+#endif
 		fprintf(stderr, "  this update: %s",
 		    t != -1 ? ctime(&t) : "\n");
 		t = tls_peer_ocsp_next_update(tls_ctx);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		add_leap_seconds(&t);
+#endif
 		fprintf(stderr, "  next update: %s",
 		    t != -1 ? ctime(&t) : "\n");
 		t = tls_peer_ocsp_revocation_time(tls_ctx);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		add_leap_seconds(&t);
+#endif
 		fprintf(stderr, "  revocation: %s",
 		    t != -1 ? ctime(&t) : "\n");
 		break;
diff -Paur --no-dereference -- libssl.upstream/apps/openssl/apps.c libssl/apps/openssl/apps.c
--- libssl.upstream/apps/openssl/apps.c
+++ libssl/apps/openssl/apps.c
@@ -147,6 +147,11 @@
 #include <openssl/x509.h>
 #include <openssl/x509v3.h>
 
+/* PATCH: For snprintf path creation that handles too long paths. */
+#if defined(__sortix__) && !defined(PATH_MAX)
+#define PATH_MAX 4096
+#endif
+
 typedef struct {
 	const char *name;
 	unsigned long flag;
diff -Paur --no-dereference -- libssl.upstream/apps/openssl/ca.c libssl/apps/openssl/ca.c
--- libssl.upstream/apps/openssl/ca.c
+++ libssl/apps/openssl/ca.c
@@ -81,6 +81,13 @@
 #include <openssl/x509.h>
 #include <openssl/x509v3.h>
 
+/* PATCH: Sortix doesn't have a PATH_MAX limit. Provide a simple value for the
+          below snprintf calls that construct paths and handle overflow if paths
+          are longer than PATH_MAX. */
+#if defined(__sortix__) && !defined(PATH_MAX)
+#define PATH_MAX 4096
+#endif
+
 #define BASE_SECTION		"ca"
 
 #define ENV_DEFAULT_CA		"default_ca"
diff -Paur --no-dereference -- libssl.upstream/apps/openssl/certhash.c libssl/apps/openssl/certhash.c
--- libssl.upstream/apps/openssl/certhash.c
+++ libssl/apps/openssl/certhash.c
@@ -33,6 +33,11 @@
 
 #include "apps.h"
 
+/* PATCH: For readlink that handles too long paths. */
+#if defined(__sortix__) && !defined(PATH_MAX)
+#define PATH_MAX 4096
+#endif
+
 static struct {
 	int dryrun;
 	int verbose;
@@ -494,6 +499,11 @@
 		fprintf(stderr, "symbolic link is too long %s\n", dep->d_name);
 		return (-1);
 	}
+	/* PATCH: Handle symbolic links that are too long. */
+	if (n == sizeof(target) - 1) {
+		fprintf(stderr, "symbolic link is too long %s\n", dep->d_name);
+		return (-1);
+	}
 	target[n] = '\0';
 
 	hi = hashinfo_from_linkname(dep->d_name, target);
diff -Paur --no-dereference -- libssl.upstream/apps/openssl/s_socket.c libssl/apps/openssl/s_socket.c
--- libssl.upstream/apps/openssl/s_socket.c
+++ libssl/apps/openssl/s_socket.c
@@ -73,6 +73,23 @@
 
 #include "s_apps.h"
 
+/* PATCH: The server below only supports IPv4 and only uses the obsolete
+          gethostbyname and gethostbyaddr instead of the getaddrinfo
+          replacement. Sortix intentionally does not have gethostbyname, so
+          simply disable the feature for now. */
+#if defined(__sortix__)
+struct hostent
+{
+	char   *h_name;
+	char  **h_aliases;
+	int     h_addrtype;
+	int     h_length;
+	char  **h_addr_list;
+};
+#define gethostbyname(a) ((void) a, (struct hostent*) NULL)
+#define gethostbyaddr(a, b, c) (((void) a), ((void) b), ((void) c), (struct hostent*) NULL)
+#endif
+
 static int init_server(int *sock, int port, int type);
 static int init_server_long(int *sock, int port, char *ip, int type);
 static int do_accept(int acc_sock, int *sock);
diff -Paur --no-dereference -- libssl.upstream/apps/openssl/x509.c libssl/apps/openssl/x509.c
--- libssl.upstream/apps/openssl/x509.c
+++ libssl/apps/openssl/x509.c
@@ -1275,7 +1275,11 @@
 		}
 	}
 	if (cfg.checkend) {
-		time_t tcheck = time(NULL) + cfg.checkoffset;
+		time_t now = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		sub_leap_seconds(&now);
+#endif
+		time_t tcheck = now + cfg.checkoffset;
 		int timecheck = X509_cmp_time(X509_get_notAfter(x), &tcheck);
 		if (timecheck == 0) {
 			BIO_printf(out, "Certificate expiry time is invalid\n");
diff -Paur --no-dereference -- libssl.upstream/configure libssl/configure
--- libssl.upstream/configure
+++ libssl/configure
@@ -11779,6 +11779,10 @@
 		HOST_OS=midipix
 		CPPFLAGS="$CPPFLAGS -D_DEFAULT_SOURCE -D_BSD_SOURCE -D_POSIX_SOURCE -D_GNU_SOURCE"
 		;;
+	*sortix*)
+		HOST_OS=sortix
+		HOST_ABI=elf
+		;;
 	*netbsd*)
 		HOST_OS=netbsd
 		HOST_ABI=elf
diff -Paur --no-dereference -- libssl.upstream/crypto/asn1/a_time.c libssl/crypto/asn1/a_time.c
--- libssl.upstream/crypto/asn1/a_time.c
+++ libssl/crypto/asn1/a_time.c
@@ -100,6 +100,9 @@
 		return ASN1_time_parse(s->data, s->length, tm, 0) != -1;
 
 	time(&now);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&now);
+#endif
 	memset(tm, 0, sizeof(*tm));
 
 	return asn1_time_time_t_to_tm(&now, tm);
diff -Paur --no-dereference -- libssl.upstream/crypto/bio/b_sock.c libssl/crypto/bio/b_sock.c
--- libssl.upstream/crypto/bio/b_sock.c
+++ libssl/crypto/bio/b_sock.c
@@ -34,6 +34,21 @@
 #include <openssl/buffer.h>
 #include <openssl/err.h>
 
+/* PATCH: Sortix intentionally doesn't have the obsolete gethostbyname family
+          but only the modern getaddrinfo(3) family. Actually none of my ports
+          use BIO_gethostbyname, maybe it can just be removed. */
+#if defined(__sortix__)
+struct hostent
+{
+	char   *h_name;
+	char  **h_aliases;
+	int     h_addrtype;
+	int     h_length;
+	char  **h_addr_list;
+};
+#define gethostbyname(a) ((void) a, (struct hostent*) NULL)
+#endif
+
 int
 BIO_get_host_ip(const char *str, unsigned char *ip)
 {
diff -Paur --no-dereference -- libssl.upstream/crypto/bio/bss_bio.c libssl/crypto/bio/bss_bio.c
--- libssl.upstream/crypto/bio/bss_bio.c
+++ libssl/crypto/bio/bss_bio.c
@@ -74,6 +74,11 @@
 # endif
 #endif
 
+/* PATCH: Sortix <limits.h> is buggy and doesn't define SSIZE_MAX right now. */
+#if defined(__sortix__)
+#include <sys/types.h>
+#endif
+
 #include <assert.h>
 #include <limits.h>
 #include <stdlib.h>
diff -Paur --no-dereference -- libssl.upstream/crypto/ct/ct_policy.c libssl/crypto/ct/ct_policy.c
--- libssl.upstream/crypto/ct/ct_policy.c
+++ libssl/crypto/ct/ct_policy.c
@@ -81,7 +81,11 @@
 	}
 
 	/* time(NULL) shouldn't ever fail, so don't bother checking for -1. */
-	ctx->epoch_time_in_ms = (uint64_t)(time(NULL) + SCT_CLOCK_DRIFT_TOLERANCE) *
+	time_t now = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&now);
+#endif
+	ctx->epoch_time_in_ms = (uint64_t)(now + SCT_CLOCK_DRIFT_TOLERANCE) *
             1000;
 
 	return ctx;
diff -Paur --no-dereference -- libssl.upstream/crypto/ocsp/ocsp_srv.c libssl/crypto/ocsp/ocsp_srv.c
--- libssl.upstream/crypto/ocsp/ocsp_srv.c
+++ libssl/crypto/ocsp/ocsp_srv.c
@@ -259,8 +259,12 @@
 		rid->type = V_OCSP_RESPID_NAME;
 	}
 
+	time_t now = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&now);
+#endif
 	if (!(flags & OCSP_NOTIME) &&
-	    !ASN1_GENERALIZEDTIME_set(brsp->tbsResponseData->producedAt, time(NULL)))
+	    !ASN1_GENERALIZEDTIME_set(brsp->tbsResponseData->producedAt, now))
 		goto err;
 
 	/* Right now, I think that not doing double hashing is the right
diff -Paur --no-dereference -- libssl.upstream/crypto/ts/ts_rsp_sign.c libssl/crypto/ts/ts_rsp_sign.c
--- libssl.upstream/crypto/ts/ts_rsp_sign.c
+++ libssl/crypto/ts/ts_rsp_sign.c
@@ -126,6 +126,9 @@
 		TS_RESP_CTX_add_failure_info(ctx, TS_INFO_TIME_NOT_AVAILABLE);
 		return 0;
 	}
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&tv.tv_sec);
+#endif
 	/* Return time to caller. */
 	*sec = tv.tv_sec;
 	*usec = tv.tv_usec;
@@ -976,6 +979,9 @@
 	if (precision > TS_MAX_CLOCK_PRECISION_DIGITS)
 		goto err;
 
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	add_leap_seconds(&sec);
+#endif
 	if (!(tm = gmtime(&sec)))
 		goto err;
 
diff -Paur --no-dereference -- libssl.upstream/crypto/x509/x509_verify.c libssl/crypto/x509/x509_verify.c
--- libssl.upstream/crypto/x509/x509_verify.c
+++ libssl/crypto/x509/x509_verify.c
@@ -75,7 +75,13 @@
 	 * a time_t. A time_t must be sane if you care about times after
 	 * Jan 19 2038.
 	 */
-	return timegm(&tm);
+	time_t secs = timegm(&tm);
+	if (secs < 0)
+		return -1;
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&secs);
+#endif
+	return secs;
 }
 
 /*
@@ -816,9 +822,12 @@
 {
 	time_t when;
 
-	if (cmp_time == NULL)
+	if (cmp_time == NULL) {
 		when = time(NULL);
-	else
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		sub_leap_seconds(&when);
+#endif
+	} else
 		when = *cmp_time;
 
 	if (cert->not_before == -1) {
diff -Paur --no-dereference -- libssl.upstream/crypto/x509/x509_vfy.c libssl/crypto/x509/x509_vfy.c
--- libssl.upstream/crypto/x509/x509_vfy.c
+++ libssl/crypto/x509/x509_vfy.c
@@ -1881,8 +1881,12 @@
 		ptime = ctx->param->check_time;
 	else if (ctx->param->flags & X509_V_FLAG_NO_CHECK_TIME)
 		return 1;
-	else
+	else {
 		ptime = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		sub_leap_seconds(&ptime);
+#endif
+	}
 
 	if (x->ex_flags & EXFLAG_SET)
 		i = time_t_bogocmp(x->not_before, ptime);
@@ -2032,9 +2036,12 @@
 {
 	time_t compare, cert_time;
 
-	if (cmp_time == NULL)
+	if (cmp_time == NULL) {
 		compare = time(NULL);
-	else
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		sub_leap_seconds(&compare);
+#endif
+	} else
 		compare = *cmp_time;
 
 	if ((cert_time = x509_verify_asn1_time_to_time_t(ctm, is_notafter)) ==
@@ -2073,9 +2080,12 @@
 X509_time_adj_ex(ASN1_TIME *s, int offset_day, long offset_sec, time_t *in_time)
 {
 	time_t t;
-	if (in_time == NULL)
+	if (in_time == NULL) {
 		t = time(NULL);
-	else
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+		sub_leap_seconds(&t);
+#endif
+	} else
 		t = *in_time;
 
 	return ASN1_TIME_adj(s, t, offset_day, offset_sec);
diff -Paur --no-dereference -- libssl.upstream/include/compat/stdlib.h libssl/include/compat/stdlib.h
--- libssl.upstream/include/compat/stdlib.h
+++ libssl/include/compat/stdlib.h
@@ -16,7 +16,7 @@
 #ifndef LIBCRYPTOCOMPAT_STDLIB_H
 #define LIBCRYPTOCOMPAT_STDLIB_H
 
-#include <sys/types.h>
+/* PATCH: <sys/types.h> doesn't need to be included. */
 #include <stdint.h>
 
 #ifndef HAVE_ARC4RANDOM_BUF
diff -Paur --no-dereference -- libssl.upstream/include/compat/string.h libssl/include/compat/string.h
--- libssl.upstream/include/compat/string.h
+++ libssl/include/compat/string.h
@@ -16,7 +16,7 @@
 #include_next <string.h>
 #endif
 
-#include <sys/types.h>
+/* PATCH: <sys/types.h> doesn't need to be included. */
 
 #if defined(__sun) || defined(_AIX) || defined(__hpux)
 /* Some functions historically defined in string.h were placed in strings.h by
diff -Paur --no-dereference -- libssl.upstream/include/compat/sys/types.h libssl/include/compat/sys/types.h
--- libssl.upstream/include/compat/sys/types.h
+++ libssl/include/compat/sys/types.h
@@ -16,7 +16,7 @@
 #ifndef LIBCRYPTOCOMPAT_SYS_TYPES_H
 #define LIBCRYPTOCOMPAT_SYS_TYPES_H
 
-#include <stdint.h>
+/* PATCH: <stdint.h> doesn't need to be included. */
 
 #ifdef __MINGW32__
 #include <_bsd_types.h>
diff -Paur --no-dereference -- libssl.upstream/include/compat/unistd.h libssl/include/compat/unistd.h
--- libssl.upstream/include/compat/unistd.h
+++ libssl/include/compat/unistd.h
@@ -64,9 +64,10 @@
 #endif
 #endif
 
-#ifndef HAVE_GETPAGESIZE
-int getpagesize(void);
-#endif
+/* PATCH: Somehow HAVE_GETPAGESIZE doesn't get defined even though configure
+          does detect Sortix has it (with the fix). Sortix getpagesize returns
+          size_t rather than int, so remove this conflicting forward
+          declaration. */
 
 #define pledge(request, paths) 0
 #define unveil(path, permissions) 0
diff -Paur --no-dereference -- libssl.upstream/ssl/ssl_asn1.c libssl/ssl/ssl_asn1.c
--- libssl.upstream/ssl/ssl_asn1.c
+++ libssl/ssl/ssl_asn1.c
@@ -307,6 +307,9 @@
 
 	/* Time [1]. */
 	s->time = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&s->time);
+#endif
 	if (!CBS_get_optional_asn1_uint64(&session, &stime, SSLASN1_TIME_TAG,
 	    0))
 		goto err;
diff -Paur --no-dereference -- libssl.upstream/ssl/ssl_lib.c libssl/ssl/ssl_lib.c
--- libssl.upstream/ssl/ssl_lib.c
+++ libssl/ssl/ssl_lib.c
@@ -2461,8 +2461,13 @@
 			connections = s->session_ctx->stats.sess_connect_good;
 		else
 			connections = s->session_ctx->stats.sess_accept_good;
-		if ((connections & 0xff) == 0xff)
-			SSL_CTX_flush_sessions(s->session_ctx, time(NULL));
+		if ((connections & 0xff) == 0xff) {
+			time_t now = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+			sub_leap_seconds(&now);
+#endif
+			SSL_CTX_flush_sessions(s->session_ctx, now);
+		}
 	}
 }
 
diff -Paur --no-dereference -- libssl.upstream/ssl/tls13_lib.c libssl/ssl/tls13_lib.c
--- libssl.upstream/ssl/tls13_lib.c
+++ libssl/ssl/tls13_lib.c
@@ -393,6 +393,9 @@
 		goto err;
 
 	sess->time = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&sess->time);
+#endif
 
 	sess->tlsext_tick_lifetime_hint = ticket_lifetime;
 	sess->tlsext_tick_age_add = ticket_age_add;
diff -Paur --no-dereference -- libssl.upstream/tls/tls_config.c libssl/tls/tls_config.c
--- libssl.upstream/tls/tls_config.c
+++ libssl/tls/tls_config.c
@@ -721,8 +721,11 @@
 	}
 
 	if (sb.st_uid != getuid()) {
+		/* PATCH: Sortix has 64-bit uid_t. */
 		tls_config_set_errorx(config, "session file has incorrect "
-		    "owner (uid %u != %u)", sb.st_uid, getuid());
+		    "owner (uid %llu != %llu)",
+		    (unsigned long long) sb.st_uid,
+		    (unsigned long long) getuid());
 		return (-1);
 	}
 	mugo = sb.st_mode & (S_IRWXU|S_IRWXG|S_IRWXO);
@@ -890,6 +893,9 @@
 	memcpy(newkey.hmac_key, key + sizeof(newkey.aes_key),
 	    sizeof(newkey.hmac_key));
 	newkey.time = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&newkey.time);
+#endif
 
 	for (i = 0; i < TLS_NUM_TICKETS; i++) {
 		struct tls_ticket_key *tk = &config->ticket_keys[i];
diff -Paur --no-dereference -- libssl.upstream/tls/tls_conninfo.c libssl/tls/tls_conninfo.c
--- libssl.upstream/tls/tls_conninfo.c
+++ libssl/tls/tls_conninfo.c
@@ -126,6 +126,10 @@
 		goto err;
 	if ((*notafter = timegm(&after_tm)) == -1)
 		goto err;
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(notbefore);
+	sub_leap_seconds(notafter);
+#endif
 
 	return (0);
 
diff -Paur --no-dereference -- libssl.upstream/tls/tls_ocsp.c libssl/tls/tls_ocsp.c
--- libssl.upstream/tls/tls_ocsp.c
+++ libssl/tls/tls_ocsp.c
@@ -67,6 +67,9 @@
 		return -1;
 	if ((*gt_time = timegm(&tm)) == -1)
 		return -1;
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(gt_time);
+#endif
 	return 0;
 }
 
diff -Paur --no-dereference -- libssl.upstream/tls/tls_server.c libssl/tls/tls_server.c
--- libssl.upstream/tls/tls_server.c
+++ libssl/tls/tls_server.c
@@ -145,6 +145,9 @@
 	int i;
 
 	now = time(NULL);
+#ifdef CLOCK_REALTIME_HAS_LEAP_SECONDS
+	sub_leap_seconds(&now);
+#endif
 	if (config->ticket_autorekey == 1) {
 		if (now - 3 * (config->session_lifetime / 4) >
 		    config->ticket_keys[0].time) {
