diff -Paur --no-dereference -- libffi.upstream/configure libffi/configure
--- libffi.upstream/configure
+++ libffi/configure
@@ -22146,7 +22146,7 @@
       eval "${MAKE-make} -f $tmp/conftemp.mk 2>/dev/null >/dev/null"
       if grep '^REMOVED,,,' "$ac_top_srcdir/Makefile" >/dev/null
       then rm $ac_top_srcdir/Makefile ; fi
-      cp $tmp/conftemp.mk $ac_top_srcdir/makefiles.mk~      ## DEBUGGING
+      #cp $tmp/conftemp.mk $ac_top_srcdir/makefiles.mk~      ## DEBUGGING
     fi
     if test ! -f "$ac_top_srcdir/Makefile" ; then
       { printf "%s\n" "$as_me:${as_lineno-$LINENO}: create top_srcdir/Makefile guessed from local Makefile" >&5
@@ -22231,7 +22231,7 @@
 	; echo "\$\$i" | grep "^../" > /dev/null && continue \\\\\\
 	; echo "rm -r \$\$i"; (rm -r "\$\$i") ; done ; rm Makefile
 _EOF
-      cp "$tmp/conftemp.sed" "$ac_top_srcdir/makefile.sed~"            ## DEBUGGING
+      #cp "$tmp/conftemp.sed" "$ac_top_srcdir/makefile.sed~"            ## DEBUGGING
       $ax_enable_builddir_sed -f $tmp/conftemp.sed Makefile >$ac_top_srcdir/Makefile
       if test -f "$ac_top_srcdir/Makefile.mk" ; then
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: extend top_srcdir/Makefile with top_srcdir/Makefile.mk" >&5
@@ -22252,8 +22252,8 @@
       # echo "/^$xxxx *$ax_enable_builddir_host /d" >$tmp/conftemp.sed
       echo "s!^$xxxx [^|]* | *$ax_enable_builddir *\$!$xxxx ...... $ax_enable_builddir!" >$tmp/conftemp.sed
       $ax_enable_builddir_sed -f "$tmp/conftemp.sed" "$ac_top_srcdir/Makefile" >$tmp/mkfile.tmp
-        cp "$tmp/conftemp.sed" "$ac_top_srcdir/makefiles.sed~"         ## DEBUGGING
-        cp "$tmp/mkfile.tmp"   "$ac_top_srcdir/makefiles.out~"         ## DEBUGGING
+        #cp "$tmp/conftemp.sed" "$ac_top_srcdir/makefiles.sed~"         ## DEBUGGING
+        #cp "$tmp/mkfile.tmp"   "$ac_top_srcdir/makefiles.out~"         ## DEBUGGING
       if cmp -s "$ac_top_srcdir/Makefile" "$tmp/mkfile.tmp" 2>/dev/null ; then
         { printf "%s\n" "$as_me:${as_lineno-$LINENO}: keeping top_srcdir/Makefile from earlier configure" >&5
 printf "%s\n" "$as_me: keeping top_srcdir/Makefile from earlier configure" >&6;}
