diff -Paur --no-dereference -- xz.upstream/configure xz/configure
--- xz.upstream/configure
+++ xz/configure
@@ -17845,7 +17845,7 @@
 
 
 
-    use_additional=yes
+    use_additional=no
 
   acl_save_prefix="$prefix"
   prefix="$acl_final_prefix"
@@ -18798,7 +18798,7 @@
 
 
 
-    use_additional=yes
+    use_additional=no
 
   acl_save_prefix="$prefix"
   prefix="$acl_final_prefix"
@@ -20750,6 +20750,7 @@
 
 #if defined(_WIN32) || defined(__CYGWIN__) || defined(__OS2__) \
 		|| defined(__DJGPP__) || defined(__VMS) \
+		|| defined(__sortix__) \
 		|| defined(AMIGA) || defined(__AROS__) || defined(__QNX__)
 int main(void) { return 0; }
 #else
diff -Paur --no-dereference -- xz.upstream/src/common/tuklib_physmem.c xz/src/common/tuklib_physmem.c
--- xz.upstream/src/common/tuklib_physmem.c
+++ xz/src/common/tuklib_physmem.c
@@ -70,6 +70,11 @@
 // This sysinfo() is Linux-specific.
 #elif defined(TUKLIB_PHYSMEM_SYSINFO)
 #	include <sys/sysinfo.h>
+
+// Sortix
+#elif defined(__sortix__)
+#	include <memusage.h>
+
 #endif
 
 
@@ -228,6 +233,10 @@
 	struct sysinfo si;
 	if (sysinfo(&si) == 0)
 		ret = (uint64_t)si.totalram * si.mem_unit;
+#elif defined(__sortix__)
+	size_t statistic = MEMUSAGE_TOTAL;
+	memusage(&statistic, &statistic, 1);
+	ret = statistic;
 #endif
 
 	return ret;
