diff -Paur --no-dereference -- libcurl.upstream/Makefile.in libcurl/Makefile.in
--- libcurl.upstream/Makefile.in
+++ libcurl/Makefile.in
@@ -621,7 +621,7 @@
  $(VC14_20_LIBVCXPROJ) $(VC14_20_SRCVCXPROJ)       \
  $(VC14_30_LIBVCXPROJ) $(VC14_30_SRCVCXPROJ)
 
-bin_SCRIPTS = curl-config
+bin_SCRIPTS =
 SUBDIRS = lib docs src scripts
 DIST_SUBDIRS = $(SUBDIRS) tests packages scripts include docs
 pkgconfigdir = $(libdir)/pkgconfig
@@ -1570,6 +1570,17 @@
 distclean: distclean-recursive
 	-rm -f $(am__CONFIG_DISTCLEAN_FILES)
 	-rm -f Makefile
+	-rm -f docs/cmdline-opts/Makefile
+	-rm -rf docs/examples/.deps
+	-rm -f docs/examples/Makefile
+	-rm -f docs/libcurl/Makefile
+	-rm -f docs/libcurl/opts/Makefile
+	-rm -f docs/Makefile
+	-rm -f docs/cmdline-opts/curl.1
+	-rm -f docs/cmdline-opts/curl.txt
+	-rm -f docs/libcurl/*.3
+	-rm -f docs/libcurl/opts/*.3
+	-rm -f scripts/Makefile
 distclean-am: clean-am distclean-generic distclean-libtool \
 	distclean-tags
 
diff -Paur --no-dereference -- libcurl.upstream/configure libcurl/configure
--- libcurl.upstream/configure
+++ libcurl/configure
@@ -23081,9 +23081,7 @@
 
 fi
 
-if test "$HAVE_GETHOSTBYNAME" != "1"; then
-  as_fn_error $? "couldn't find libraries for gethostbyname()" "$LINENO" 5
-fi
+# PATCH: gethostbyname is not actually needed.
 
 
 curl_includes_winsock2="\
@@ -37811,11 +37809,8 @@
 
 int main (void)
 {
-#ifdef h_errno
+  /* PATCH: h_errno has nothing to do with getaddrinfo */
   return 0;
-#else
-  #error force compilation error
-#endif
 }
 
 
diff -Paur --no-dereference -- libcurl.upstream/docs/Makefile.in libcurl/docs/Makefile.in
--- libcurl.upstream/docs/Makefile.in
+++ libcurl/docs/Makefile.in
@@ -487,8 +487,8 @@
 
 # if we disable man page building, ignore these
 @BUILD_DOCS_TRUE@MK_CA_DOCS = mk-ca-bundle.1
-@BUILD_DOCS_TRUE@CURLCONF_DOCS = curl-config.1
-@BUILD_DOCS_TRUE@man_MANS = curl-config.1
+@BUILD_DOCS_TRUE@CURLCONF_DOCS =
+@BUILD_DOCS_TRUE@man_MANS = 
 CURLPAGES = curl-config.md mk-ca-bundle.md
 SUBDIRS = . cmdline-opts libcurl
 DIST_SUBDIRS = $(SUBDIRS) examples
diff -Paur --no-dereference -- libcurl.upstream/include/curl/curl.h libcurl/include/curl/curl.h
--- libcurl.upstream/include/curl/curl.h
+++ libcurl/include/curl/curl.h
@@ -83,18 +83,7 @@
 #endif
 #endif
 
-/* HP-UX systems version 9, 10 and 11 lack sys/select.h and so does oldish
-   libc5-based Linux systems. Only include it on systems that are known to
-   require it! */
-#if defined(_AIX) || defined(__NOVELL_LIBC__) || defined(__NetBSD__) || \
-    defined(__minix) || defined(__INTEGRITY) || \
-    defined(ANDROID) || defined(__ANDROID__) || defined(__OpenBSD__) || \
-    defined(__CYGWIN__) || defined(AMIGA) || defined(__NuttX__) || \
-   (defined(__FreeBSD_version) && (__FreeBSD_version < 800000)) || \
-   (defined(__MidnightBSD_version) && (__MidnightBSD_version < 100000)) || \
-    defined(__sun__) || defined(__serenity__) || defined(__vxworks__)
 #include <sys/select.h>
-#endif
 
 #if !defined(_WIN32) && !defined(_WIN32_WCE)
 #include <sys/socket.h>
diff -Paur --no-dereference -- libcurl.upstream/lib/curl_addrinfo.h libcurl/lib/curl_addrinfo.h
--- libcurl.upstream/lib/curl_addrinfo.h
+++ libcurl/lib/curl_addrinfo.h
@@ -71,6 +71,16 @@
                     struct Curl_addrinfo **result);
 #endif
 
+#if defined(__sortix__)
+struct hostent
+{
+   char* h_name;
+   char** h_aliases;
+   int h_addrtype;
+   int h_length;
+   char** h_addr_list;
+};
+#endif
 struct Curl_addrinfo *
 Curl_he2ai(const struct hostent *he, int port);
 
