diff -Paur --no-dereference -- libcurl.upstream/Makefile.in libcurl/Makefile.in
--- libcurl.upstream/Makefile.in
+++ libcurl/Makefile.in
@@ -659,7 +659,7 @@
  $(VC14_SRCVCXPROJ) $(VC14_10_LIBVCXPROJ) $(VC14_10_SRCVCXPROJ)              \
  $(VC14_30_LIBVCXPROJ) $(VC14_30_SRCVCXPROJ)
 
-bin_SCRIPTS = curl-config
+bin_SCRIPTS =
 SUBDIRS = lib src
 DIST_SUBDIRS = $(SUBDIRS) tests packages scripts include docs
 pkgconfigdir = $(libdir)/pkgconfig
@@ -1593,6 +1593,12 @@
 distclean: distclean-recursive
 	-rm -f $(am__CONFIG_DISTCLEAN_FILES)
 	-rm -f Makefile
+	-rm -f docs/cmdline-opts/Makefile
+	-rm -rf docs/examples/.deps
+	-rm -f docs/examples/Makefile
+	-rm -f docs/libcurl/Makefile
+	-rm -f docs/libcurl/opts/Makefile
+	-rm -f docs/Makefile
 distclean-am: clean-am distclean-generic distclean-libtool \
 	distclean-tags
 
diff -Paur --no-dereference -- libcurl.upstream/configure libcurl/configure
--- libcurl.upstream/configure
+++ libcurl/configure
@@ -23303,9 +23303,7 @@
 
 fi
 
-if test "$HAVE_GETHOSTBYNAME" != "1"; then
-  as_fn_error $? "couldn't find libraries for gethostbyname()" "$LINENO" 5
-fi
+# PATCH: gethostbyname is not actually needed.
 
 
 curl_includes_winsock2="\
@@ -37520,11 +37518,8 @@
 
 int main (void)
 {
-#ifdef h_errno
+  /* PATCH: h_errno has nothing to do with getaddrinfo */
   return 0;
-#else
-  force compilation error
-#endif
 }
 
 
diff -Paur --no-dereference -- libcurl.upstream/docs/Makefile.in libcurl/docs/Makefile.in
--- libcurl.upstream/docs/Makefile.in
+++ libcurl/docs/Makefile.in
@@ -484,10 +484,10 @@
 # but distribute it (using the relative file name) in the next variable
 man_MANS = $(abs_builddir)/curl.1
 noinst_man_MANS = curl.1 mk-ca-bundle.1
-dist_man_MANS = curl-config.1
-GENHTMLPAGES = curl.html curl-config.html mk-ca-bundle.html
-PDFPAGES = curl.pdf curl-config.pdf mk-ca-bundle.pdf
-MANDISTPAGES = curl.1.dist curl-config.1.dist
+dist_man_MANS =
+GENHTMLPAGES = curl.html mk-ca-bundle.html
+PDFPAGES = curl.pdf mk-ca-bundle.pdf
+MANDISTPAGES = curl.1.dist
 HTMLPAGES = $(GENHTMLPAGES)
 
 # Build targets in this file (.) before cmdline-opts to ensure that
diff -Paur --no-dereference -- libcurl.upstream/include/curl/curl.h libcurl/include/curl/curl.h
--- libcurl.upstream/include/curl/curl.h
+++ libcurl/include/curl/curl.h
@@ -75,18 +75,7 @@
 #endif
 #endif
 
-/* HP-UX systems version 9, 10 and 11 lack sys/select.h and so does oldish
-   libc5-based Linux systems. Only include it on systems that are known to
-   require it! */
-#if defined(_AIX) || defined(__NOVELL_LIBC__) || defined(__NetBSD__) || \
-    defined(__minix) || defined(__INTEGRITY) || \
-    defined(ANDROID) || defined(__ANDROID__) || defined(__OpenBSD__) || \
-    defined(__CYGWIN__) || defined(AMIGA) || defined(__NuttX__) || \
-   (defined(__FreeBSD_version) && (__FreeBSD_version < 800000)) || \
-   (defined(__MidnightBSD_version) && (__MidnightBSD_version < 100000)) || \
-    defined(__sun__) || defined(__serenity__) || defined(__vxworks__)
 #include <sys/select.h>
-#endif
 
 #if !defined(_WIN32) && !defined(_WIN32_WCE)
 #include <sys/socket.h>
diff -Paur --no-dereference -- libcurl.upstream/lib/curl_addrinfo.h libcurl/lib/curl_addrinfo.h
--- libcurl.upstream/lib/curl_addrinfo.h
+++ libcurl/lib/curl_addrinfo.h
@@ -71,6 +71,16 @@
                     struct Curl_addrinfo **result);
 #endif
 
+#if defined(__sortix__)
+struct hostent
+{
+   char* h_name;
+   char** h_aliases;
+   int h_addrtype;
+   int h_length;
+   char** h_addr_list;
+};
+#endif
 struct Curl_addrinfo *
 Curl_he2ai(const struct hostent *he, int port);
 
