diff -Paur --no-dereference -- gawk.upstream/extension/filefuncs.c gawk/extension/filefuncs.c
--- gawk.upstream/extension/filefuncs.c
+++ gawk/extension/filefuncs.c
@@ -418,8 +418,12 @@
 	/* for block and character devices, add rdev, major and minor numbers */
 	if (S_ISBLK(sbuf->st_mode) || S_ISCHR(sbuf->st_mode)) {
 		array_set_numeric(array, "rdev", sbuf->st_rdev);
+#ifdef major
 		array_set_numeric(array, "major", major(sbuf->st_rdev));
+#endif
+#ifdef minor
 		array_set_numeric(array, "minor", minor(sbuf->st_rdev));
+#endif
 	}
 
 #ifdef HAVE_STRUCT_STAT_ST_BLKSIZE
diff -Paur --no-dereference -- gawk.upstream/extras/Makefile.in gawk/extras/Makefile.in
--- gawk.upstream/extras/Makefile.in
+++ gawk/extras/Makefile.in
@@ -316,7 +316,7 @@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-profiledir = $(sysconfdir)/profile.d
+profiledir = $(sysconfdir)/default/profile.d
 profile_DATA = gawk.sh gawk.csh
 EXTRA_DIST = $(profile_DATA)
 all: all-am
diff -Paur --no-dereference -- gawk.upstream/printf.c gawk/printf.c
--- gawk.upstream/printf.c
+++ gawk/printf.c
@@ -924,7 +924,13 @@
 		 * Use snprintf return value to tell if there
 		 * is enough room in the buffer or not.
 		 */
+/* PATCH: Until Sortix has floating point printing. */
+#if defined(__sortix__) && !__SORTIX_HAS_FLOAT_PARSE__
+		while ((i = snprintf(buf, buflen, "%lld", (long long)tmpval)) >=
+			buflen) {
+#else
 		while ((i = snprintf(buf, buflen, "%.0f", tmpval)) >= buflen) {
+#endif
 			if (i > 0)
 				buflen += (i > buflen) ?  i : buflen;
 			else
diff -Paur --no-dereference -- gawk.upstream/main.c gawk/main.c
--- gawk.upstream/main.c
+++ gawk/main.c
@@ -1028,13 +1028,7 @@
 	update_PROCINFO_num("api_minor", GAWK_API_MINOR_VERSION);
 #endif
 
-#ifdef GETPGRP_VOID
-#define getpgrp_arg() /* nothing */
-#else
-#define getpgrp_arg() getpid()
-#endif
-
-	value = getpgrp(getpgrp_arg());
+	value = getpgid(0);
 	update_PROCINFO_num("pgrpid", value);
 
 	/*
diff -Paur --no-dereference -- gawk.upstream/node.c gawk/node.c
--- gawk.upstream/node.c
+++ gawk/node.c
@@ -1020,6 +1020,8 @@
 	n->flags &= ~WSTRCUR;
 }
 
+/* PATCH: Sortix doesn't have putwc(3) and this function is unused. */
+#if 0
 static void __attribute__ ((unused))
 dump_wstr(FILE *fp, const wchar_t *str, size_t len)
 {
@@ -1029,6 +1031,7 @@
 	for (; len--; str++)
 		putwc(*str, fp);
 }
+#endif
 
 /* wstrstr --- walk haystack, looking for needle, wide char version */
 
