diff -Paur --no-dereference -- libreadline.upstream/Makefile.in libreadline/Makefile.in
--- libreadline.upstream/Makefile.in
+++ libreadline/Makefile.in
@@ -67,9 +67,6 @@
 
 man3dir = $(mandir)/man3
 
-# Support an alternate destination root directory for package building
-DESTDIR =
-
 # Programs to make tags files.
 ETAGS = etags
 CTAGS = ctags -w
diff -Paur --no-dereference -- libreadline.upstream/configure libreadline/configure
--- libreadline.upstream/configure
+++ libreadline/configure
@@ -3075,8 +3075,6 @@
     i[3456]86-*-beos*)
         cross_cache=${srcdir}/cross-build/x86-beos.cache
         ;;
-    *)  echo "configure: cross-compiling for $host is not supported" >&2
-        ;;
     esac
     if test -n "${cross_cache}" && test -r "${cross_cache}"; then
         echo "loading cross-build cache file ${cross_cache}"
diff -Paur --no-dereference -- libreadline.upstream/input.c libreadline/input.c
--- libreadline.upstream/input.c
+++ libreadline/input.c
@@ -805,7 +805,7 @@
   int result;
   unsigned char c;
   int fd;
-#if defined (HAVE_PSELECT)
+#if defined (HAVE_PSELECT) || defined (HAVE_SELECT)
   sigset_t empty_set;
   fd_set readfds;
 #endif
diff -Paur --no-dereference -- libreadline.upstream/terminal.c libreadline/terminal.c
--- libreadline.upstream/terminal.c
+++ libreadline/terminal.c
@@ -102,13 +102,6 @@
 
 static int tcap_initialized;
 
-#if !defined (__linux__) && !defined (NCURSES_VERSION)
-#  if defined (__EMX__) || defined (NEED_EXTERN_PC)
-extern 
-#  endif /* __EMX__ || NEED_EXTERN_PC */
-char PC, *BC, *UP;
-#endif /* !__linux__ && !NCURSES_VERSION */
-
 /* Some strings to control terminal actions.  These are output by tputs (). */
 char *_rl_term_clreol;
 char *_rl_term_clrpag;
