diff -Paur --no-dereference -- libfreetype.upstream/builds/unix/configure libfreetype/builds/unix/configure
--- libfreetype.upstream/builds/unix/configure
+++ libfreetype/builds/unix/configure
@@ -14572,56 +14572,8 @@
       have_libpng="yes (LIBPNG_CFLAGS and LIBPNG_LIBS)"
     else
       # fall back to config script
-      # Extract the first word of "libpng-config", so it can be a program name with args.
-set dummy libpng-config; ac_word=$2
-{ printf "%s\n" "$as_me:${as_lineno-$LINENO}: checking for $ac_word" >&5
-printf %s "checking for $ac_word... " >&6; }
-if test ${ac_cv_prog_have_libpng+y}
-then :
-  printf %s "(cached) " >&6
-else case e in #(
-  e) if test -n "$have_libpng"; then
-  ac_cv_prog_have_libpng="$have_libpng" # Let the user override the test.
-else
-as_save_IFS=$IFS; IFS=$PATH_SEPARATOR
-for as_dir in $PATH
-do
-  IFS=$as_save_IFS
-  case $as_dir in #(((
-    '') as_dir=./ ;;
-    */) ;;
-    *) as_dir=$as_dir/ ;;
-  esac
-    for ac_exec_ext in '' $ac_executable_extensions; do
-  if as_fn_executable_p "$as_dir$ac_word$ac_exec_ext"; then
-    ac_cv_prog_have_libpng="yes (libpng-config)"
-    printf "%s\n" "$as_me:${as_lineno-$LINENO}: found $as_dir$ac_word$ac_exec_ext" >&5
-    break 2
-  fi
-done
-  done
-IFS=$as_save_IFS
-
-  test -z "$ac_cv_prog_have_libpng" && ac_cv_prog_have_libpng="no"
-fi ;;
-esac
-fi
-have_libpng=$ac_cv_prog_have_libpng
-if test -n "$have_libpng"; then
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: $have_libpng" >&5
-printf "%s\n" "$have_libpng" >&6; }
-else
-  { printf "%s\n" "$as_me:${as_lineno-$LINENO}: result: no" >&5
-printf "%s\n" "no" >&6; }
-fi
-
-
-      if test "$have_libpng" != no; then
-        LIBPNG_CFLAGS=`libpng-config --cflags`
-        LIBPNG_LIBS=`libpng-config --ldflags`
-        libpng_libspriv=`libpng-config --static --ldflags`
-        libpng_libsstaticconf="$libpng_libspriv"
-      fi
+      # PATCH: libpng-config should never be used.
+      :
     fi
   fi
 fi
