diff -Paur --no-dereference -- nasm.upstream/include/compiler.h nasm/include/compiler.h
--- nasm.upstream/include/compiler.h
+++ nasm/include/compiler.h
@@ -414,4 +414,9 @@
 # define default case BOGUS_CASE: default
 #endif
 
+/* PATCH: Sortix doesn't have FILENAME_MAX at this time. */
+#if !defined(FILENAME_MAX) && defined(__sortix__)
+#define FILENAME_MAX 256
+#endif
+
 #endif	/* NASM_COMPILER_H */
