diff -Paur --no-dereference -- perl.upstream/cnf/configure perl/cnf/configure
--- perl.upstream/cnf/configure
+++ perl/cnf/configure
@@ -48,7 +48,9 @@
 # past this point $build/$buildarch are non-empty unless we're in target mode
 
 if [ -z "$mode" ]; then
-	if [ -z "$target" -o "$target" = "$build" ]; then
+	cc_sysroot=`${CC-cc} --print-sysroot || true`
+	if [ -z "$target" -o "$target" = "$build" ] &&
+	   [ "$cc_sysroot" = "" -o "$cc_sysroot" = / ]; then
 		mode='native'
 	else
 		mode='cross'
diff -Paur --no-dereference -- perl.upstream/cnf/configure_args.sh perl/cnf/configure_args.sh
--- perl.upstream/cnf/configure_args.sh
+++ perl/cnf/configure_args.sh
@@ -91,18 +91,23 @@
 	esac
 	# split --set-foo and similar constructs into --set foo
 	# and things like --prefix=/foo into --prefix and /foo
+	# PATCH: Fix --foo= bar being parsed as --foo bar.
+	split=
 	case "$a" in
 		set-*|use-*|include-*)
 			k=`echo "$a" | sed -e 's/^[^-]*-//'`
 			a=`echo "$a" | sed -e 's/-.*//'`
+			split=1
 			;;
 		dont-use-*|dont-include-*)	
 			k=`echo "$a" | sed -e 's/^dont-[^-]*-//'`
 			a=`echo "$a" | sed -e 's/^\(dont-[^-]*\)-.*/\1/'`
+			split=1
 			;;
 		*=*)
 			k=`echo "$a" | sed -e 's/^[^=]*=//'`
 			a=`echo "$a" | sed -e 's/=.*//'`
+			split=1
 			;;
 	esac
 	# check whether kv is required
@@ -114,7 +119,7 @@
 	esac
 	# fetch argument if necessary (--set foo=bar)
 	# note that non-empty n means there must be no argument
-	if [ -n "$x" -a -z "$k" -a -z "$n" ]; then
+	if [ -z "$split" -a -n "$x" -a -z "$k" -a -z "$n" ]; then
 		eval k="\${$i}"; i=$((i+1))
 	fi
 	# split kv pair into k and v (k=foo v=bar)
@@ -155,7 +160,8 @@
 		help) mode="help" ;;
 		regen|regenerate) mode="regen" ;;
 		keeplog) defuser "$a" 1 ;;
-		prefix|html[13]dir|libsdir)	defuser $a "$v" ;;
+		# PATCH: At least recognize --exec-prefix even though it's ignored.        
+		prefix|exec-prefix|html[13]dir|libsdir)	defuser $a "$v" ;;
 		man[13]dir|otherlibsdir)	defuser $a "$v" ;;
 		siteprefix|sitehtml[13]dir)	defuser $a "$v" ;;
 		siteman[13]dir|vendorman[13]dir)defuser $a "$v" ;;
diff -Paur --no-dereference -- perl.upstream/cnf/configure_func.sh perl/cnf/configure_func.sh
--- perl.upstream/cnf/configure_func.sh
+++ perl/cnf/configure_func.sh
@@ -8,7 +8,8 @@
     try_add '#define _GNU_SOURCE'
 		funcincludes "$3" "$4" "$includes"
 		try_add "int main(void) { $2($3); return 0; }"
-		try_link -O0 -fno-builtin
+		# PATCH: Work around tests for functions without prototypes.
+		try_link -O0 -fno-builtin -Wno-error=implicit-function-declaration
 		resdef $1 'found' 'missing'
 	fi
 }
@@ -187,7 +188,8 @@
 checkfunc d_pathconf 'pathconf' 'NULL,0' 'unistd.h'
 checkfunc d_pause 'pause' '' 'unistd.h'
 checkfunc d_pipe 'pipe' 'NULL' 'fcntl.h unistd.h'
-checkfunc d_pipe2 'pipe' 'NULL,0' 'fcntl.h unistd.h'
+# PATCH: pipe2 accidentally checked for pipe.
+checkfunc d_pipe2 'pipe2' 'NULL,0' 'fcntl.h unistd.h'
 checkfunc d_poll 'poll' 'NULL,0,0' 'poll.h'
 checkfunc d_prctl 'prctl' '0,0,0,0,0' 'sys/prctl.h'
 checkfunc d_pthread_atfork 'pthread_atfork' 'NULL,NULL,NULL' 'pthread.h'
@@ -310,4 +312,6 @@
 checkfunc d_wcsxfrm 'wcsxfrm' 'NULL,NULL,0' 'wchar.h'
 checkfunc d_wctomb 'wctomb' 'NULL,NULL' 'wchar.h'
 checkfunc d_writev 'writev' '0,NULL,0' 'sys/uio.h'
+# PATCH: Add missing check for whether nanosleep exists.
+checkfunc d_nanosleep 'nanosleep' 'NULL,NULL' 'time.h'
 unset includes
diff -Paur --no-dereference -- perl.upstream/cnf/configure_func_ext.sh perl/cnf/configure_func_ext.sh
--- perl.upstream/cnf/configure_func_ext.sh
+++ perl/cnf/configure_func_ext.sh
@@ -68,7 +68,8 @@
 mstart "Checking FD_SET macros"
 if not hinted d_fd_macros 'found' 'missing'; then
 	try_start
-	try_includes sys/time.h sys/types.h unistd.h
+	# PATCH: FD_SET is defined in <sys/select.h>.
+	try_includes sys/select.h sys/time.h sys/types.h unistd.h
 	try_cat <<END
 #if defined(FD_SET) && defined(FD_CLR) && defined(FD_ISSET) && defined(FD_ZERO)
 #else
diff -Paur --no-dereference -- perl.upstream/cnf/configure_tool.sh perl/cnf/configure_tool.sh
--- perl.upstream/cnf/configure_tool.sh
+++ perl/cnf/configure_tool.sh
@@ -270,6 +270,10 @@
 			define osname "midipix"
 			result "Midipix"
 			;;
+		*-sortix*)
+			define osname "sortix"
+			result "Sortix"
+			;;
 		*)
 			result "no"
 			;;
diff -Paur --no-dereference -- perl.upstream/cnf/configure_type.sh perl/cnf/configure_type.sh
--- perl.upstream/cnf/configure_type.sh
+++ perl/cnf/configure_type.sh
@@ -102,7 +102,8 @@
 test "$d_longdbl" = 'define'  && checksize longdblsize 'long double'
 test "$d_longlong" = 'define' && checksize longlongsize 'long long'
 
-checktype d_fd_set 'fd_set' 'sys/types.h'
+# PATCH: FD_SET is defined in <sys/select.h>.
+checktype d_fd_set 'fd_set' 'sys/select.h'
 checktype d_fpos64_t 'fpos64_t' 'stdio.h'
 checktype d_off64_t 'off64_t' 'sys/types.h'
 checktype d_ptrdiff_t 'ptrdiff_t' 'stddef.h'
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/constant.applied perl/cnf/diffs/perl5-5.39.5/constant.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/constant.applied
+++ perl/cnf/diffs/perl5-5.39.5/constant.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/dynaloader.applied perl/cnf/diffs/perl5-5.39.5/dynaloader.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/dynaloader.applied
+++ perl/cnf/diffs/perl5-5.39.5/dynaloader.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/findext.applied perl/cnf/diffs/perl5-5.39.5/findext.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/findext.applied
+++ perl/cnf/diffs/perl5-5.39.5/findext.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/installscripts.applied perl/cnf/diffs/perl5-5.39.5/installscripts.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/installscripts.applied
+++ perl/cnf/diffs/perl5-5.39.5/installscripts.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/liblist.applied perl/cnf/diffs/perl5-5.39.5/liblist.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/liblist.applied
+++ perl/cnf/diffs/perl5-5.39.5/liblist.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/makemaker.applied perl/cnf/diffs/perl5-5.39.5/makemaker.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/makemaker.applied
+++ perl/cnf/diffs/perl5-5.39.5/makemaker.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/posix-makefile.applied perl/cnf/diffs/perl5-5.39.5/posix-makefile.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/posix-makefile.applied
+++ perl/cnf/diffs/perl5-5.39.5/posix-makefile.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/test-checkcase.applied perl/cnf/diffs/perl5-5.39.5/test-checkcase.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/test-checkcase.applied
+++ perl/cnf/diffs/perl5-5.39.5/test-checkcase.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/test-makemaker.applied perl/cnf/diffs/perl5-5.39.5/test-makemaker.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/test-makemaker.applied
+++ perl/cnf/diffs/perl5-5.39.5/test-makemaker.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/cnf/diffs/perl5-5.39.5/xconfig.applied perl/cnf/diffs/perl5-5.39.5/xconfig.applied
--- perl.upstream/cnf/diffs/perl5-5.39.5/xconfig.applied
+++ perl/cnf/diffs/perl5-5.39.5/xconfig.applied
@@ -0,0 +1 @@
+
diff -Paur --no-dereference -- perl.upstream/config_h.SH perl/config_h.SH
--- perl.upstream/config_h.SH
+++ perl/config_h.SH
@@ -1230,6 +1230,9 @@
  *	function used to generate normalized random numbers.
  *	Values include 15, 16, 31, and 48.
  */
+#if defined(__sortix__) && !defined(__SORTIX_HAS_DRAND48__)
+double drand48(void);
+#endif
 #define Drand01()		$drand01		/**/
 #define Rand_seed_t		$randseedtype		/**/
 #define seedDrand01(x)	$seedfunc((Rand_seed_t)x)	/**/
diff -Paur --no-dereference -- perl.upstream/cpan/Pod-Perldoc/lib/Pod/Perldoc.pm perl/cpan/Pod-Perldoc/lib/Pod/Perldoc.pm
--- perl.upstream/cpan/Pod-Perldoc/lib/Pod/Perldoc.pm
+++ perl/cpan/Pod-Perldoc/lib/Pod/Perldoc.pm
@@ -1676,7 +1676,7 @@
         if ($self->is_os2) {
           unshift @pagers, 'less', 'cmd /c more <';
         }
-        push @pagers, qw( more less pg view cat );
+        push @pagers, qw( less pager more pg view cat );
         unshift @pagers, "$ENV{PAGER} <"  if $ENV{PAGER};
     }
 
diff -Paur --no-dereference -- perl.upstream/dist/Safe/Safe.pm perl/dist/Safe/Safe.pm
--- perl.upstream/dist/Safe/Safe.pm
+++ perl/dist/Safe/Safe.pm
@@ -43,7 +43,10 @@
     }
 }
 
-use Opcode 1.01, qw(
+# PATCH: 'use Safe;' fails with '"%.*g" is not exported by the Opcode module'
+#        on Sortix without floating point support when a floating point version
+#        number is used here.
+use Opcode 1, qw(
     opset opset_to_ops opmask_add
     empty_opset full_opset invert_opset verify_opset
     opdesc opcodes opmask define_optag opset_to_hex
diff -Paur --no-dereference -- perl.upstream/ext/Errno/Errno_pm.PL perl/ext/Errno/Errno_pm.PL
--- perl.upstream/ext/Errno/Errno_pm.PL
+++ perl/ext/Errno/Errno_pm.PL
@@ -1,6 +1,9 @@
 use ExtUtils::MakeMaker;
 use Config;
 use strict;
+ 
+# PATCH: Recognize the right operating system when cross-compiling.
+$^O = $Config{osname};
 
 our $VERSION = "1.38";
 
@@ -160,6 +163,11 @@
 
 	close(CPPI);
 
+	# PATCH: Simply process errno.c rather then the special logic below to
+	# locate the system headers.
+	push(@file, "errno.c");
+	return uniq(@file);
+
 	# invoke CPP and read the output
 	if ($IsMSWin32) {
 	    open(CPPO,"$Config{cpprun} $Config{cppflags} errno.c |") or
diff -Paur --no-dereference -- perl.upstream/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm perl/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
--- perl.upstream/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
+++ perl/ext/ExtUtils-Miniperl/lib/ExtUtils/Miniperl.pm
@@ -2,7 +2,9 @@
 package ExtUtils::Miniperl;
 use strict;
 use Exporter 'import';
-use ExtUtils::Embed 1.31, qw(xsi_header xsi_protos xsi_body);
+# PATCH: Work around missing Sortix printf floating point support causing:
+#        "%.*g" is not exported by the ExtUtils::Embed module
+use ExtUtils::Embed 1, qw(xsi_header xsi_protos xsi_body);
 
 our @EXPORT = qw(writemain);
 our $VERSION = '1.14';
diff -Paur --no-dereference -- perl.upstream/ext/POSIX/POSIX.xs perl/ext/POSIX/POSIX.xs
--- perl.upstream/ext/POSIX/POSIX.xs
+++ perl/ext/POSIX/POSIX.xs
@@ -1,3 +1,12 @@
+#if defined(__sortix__) && !defined(__SORTIX_HAS_PAUSE__)
+#include <stddef.h>
+#include <signal.h>
+int pause(void)
+{
+	return sigsuspend(NULL);
+}
+#endif
+
 #define PERL_EXT_POSIX
 #define PERL_EXT
 
@@ -1418,7 +1427,11 @@
 #      define mkfifo(a,b) not_here("mkfifo")
 #    else	/* !( defined OS2 ) */
 #      ifndef mkfifo
+#if defined(__sortix__) && !defined(__SORTIX_HAS_MKFIFO__)
+#        define mkfifo(path, mode) not_here("mkfifo")
+#else
 #        define mkfifo(path, mode) (mknod((path), (mode) | S_IFIFO, 0))
+#endif
 #      endif
 #    endif
 #  endif /* !HAS_MKFIFO */
@@ -3623,12 +3636,19 @@
 ctermid(s = 0)
 	char *          s = 0;
     CODE:
+/* PATCH: Sortix doesn't have ctermid at this time. */
+#ifdef L_ctermid
 #ifdef I_TERMIOS
         /* On some systems L_ctermid is a #define; but not all; this code works
          * for all cases (so far...) */
 	s = (char *) safemalloc((size_t) L_ctermid);
 #endif
 	RETVAL = ctermid(s);
+#else
+	s = (char *) safemalloc(strlen("/dev/tty"));
+	strcpy(s, "/dev/tty");
+	RETVAL = s;
+#endif
     OUTPUT:
 	RETVAL
     CLEANUP:
diff -Paur --no-dereference -- perl.upstream/lib/unicore/mktables perl/lib/unicore/mktables
--- perl.upstream/lib/unicore/mktables
+++ perl/lib/unicore/mktables
@@ -12646,10 +12646,12 @@
         if (   defined $nv_floating_to_rational{$float}
             && $nv_floating_to_rational{$float} ne $rational)
         {
-            die Carp::my_carp_bug("Both '$rational' and"
-                            . " '$nv_floating_to_rational{$float}' evaluate to"
-                            . " the same floating point number."
-                            . "  \$E_FLOAT_PRECISION must be increased");
+            # PATCH: This test fails on native Sortix due to lack of floating
+            #        point formatting support, but nothing is wrong.
+            #die Carp::my_carp_bug("Both '$rational' and"
+            #                . " '$nv_floating_to_rational{$float}' evaluate to"
+            #                . " the same floating point number."
+            #                . "  \$E_FLOAT_PRECISION must be increased");
         }
         $nv_floating_to_rational{$float} = $rational;
     }
diff -Paur --no-dereference -- perl.upstream/Makefile perl/Makefile
--- perl.upstream/Makefile
+++ perl/Makefile
@@ -67,7 +67,7 @@
 # Original versions are not saved anymore; patch generally takes care of this,
 # and if that fails, reaching for the source tarball is the safest option.
 $(CROSSPATCHED): %.applied: %.patch
-	$(cpatch) -p1 -i $< && touch $@
+	$(cpatch) -p1 -i $< && echo > $@
 
 # ---[ common ]-----------------------------------------------------------------
 
@@ -242,6 +242,7 @@
 
 $(static_modules): %/pm_to_blib: | %/Makefile $(nonxs_tgt)
 	$(MAKE) -C $(dir $@) PERL_CORE=1 LIBPERL=$(LIBPERL) LINKTYPE=static static
+	$(MAKE) -C $(dir $@) PERL_CORE=1 LIBPERL=$(LIBPERL) LINKTYPE=static pm_to_blib
 	@touch $@
 
 $(dynamic_modules) $(disabled_dynamic): %/pm_to_blib: | %/Makefile
@@ -440,7 +442,10 @@
 
 install.sym: # deprecated
 
-install.man: installman pod/perltoc.pod | miniperl$X
+# PATCH: Sever makefile dependency that tends to livelock make with 100% CPU
+#        usage after doing the static modules when building -j2 or more.
+#        pod/perltoc.pod and miniperl are already built by the main all target.
+install.man: installman # pod/perltoc.pod | miniperl$X
 	./miniperl_top installman --destdir=$(DESTDIR) $(INSTALLFLAGS)
 
 # ---[ testpack ]---------------------------------------------------------------
@@ -453,9 +458,10 @@
 	tar -zcvf $@ -T $<
 
 # ---[ clean ]------------------------------------------------------------------
+# PATCH: make doesn't work this way when make -j.
 # clean-modules must go BEFORE clean-generated-files because it depends on config.h!
 .PHONY: clean clean-obj clean-generated-files clean-subdirs clean-modules clean-testpack
-clean: clean-obj clean-modules clean-generated-files clean-subdirs clean-testpack
+clean: clean-obj clean-modules clean-subdirs clean-testpack
 
 clean-obj:
 	-test -n "$o" && rm -f *$o
diff -Paur --no-dereference -- perl.upstream/pp.c perl/pp.c
--- perl.upstream/pp.c
+++ perl/pp.c
@@ -3196,6 +3196,13 @@
     }
 }
 
+#if defined(__sortix__) && !defined(__SORTIX_HAS_DRAND48__)
+double drand48(void)
+{
+    return arc4random() / 4294967295.0;
+}
+#endif
+
 /* Support Configure command-line overrides for rand() functions.
    After 5.005, perhaps we should replace this by Configure support
    for drand48(), random(), or rand().  For 5.005, though, maintain
diff -Paur --no-dereference -- perl.upstream/regcomp.c perl/regcomp.c
--- perl.upstream/regcomp.c
+++ perl/regcomp.c
@@ -145,6 +145,11 @@
 #include "unicode_constants.h"
 #include "regcomp_internal.h"
 
+/* PATCH: Avoid multiple definitions. */
+#ifdef PERL_EXT_RE_BUILD
+#undef PERL_RE_BUILD_AUX
+#endif
+
 /* =========================================================
  * BEGIN edit_distance stuff.
  *
diff -Paur --no-dereference -- perl.upstream/t/lib/commonsense.t perl/t/lib/commonsense.t
--- perl.upstream/t/lib/commonsense.t
+++ perl/t/lib/commonsense.t
@@ -17,7 +17,7 @@
 if (($Config{'extensions'} !~ /\bIO\s/) ){
   BAIL_OUT("Perl configured without IO module");
 }
-if (($Config{'extensions'} !~ /\bFile\/Glob\b/) ){
+if (($Config{'extensions'} !~ /\bFile.Glob\b/) ){
   BAIL_OUT("Perl configured without File::Glob module");
 }
 
diff -Paur --no-dereference -- perl.upstream/util.c perl/util.c
--- perl.upstream/util.c
+++ perl/util.c
@@ -2300,7 +2300,7 @@
  * 'current' is non-null, with up to three sizes that are added together.
  * It handles integer overflow.
  */
-#  ifndef HAS_SETENV
+/* PATCH: Support having only setenv/unsetenv and no putenv. */
 static char *
 S_env_alloc(void *current, Size_t l1, Size_t l2, Size_t l3, Size_t size)
 {
@@ -2325,7 +2325,6 @@
   panic:
     croak_memory_wrap();
 }
-#  endif
 
 /*
 =for apidoc_section $utility
diff -Paur --no-dereference -- perl.upstream/vutil.c perl/vutil.c
--- perl.upstream/vutil.c
+++ perl/vutil.c
@@ -588,6 +588,7 @@
 
 /* Macro to do the meat of getting the PV of an NV version number.  This is
  * macroized because can be called from several places */
+/* PATCH: Work around lack of Sortix printf floating point support. */
 #define GET_NUMERIC_VERSION(ver, sv, tbuf, buf, len)                        \
     STMT_START {                                                            \
                                                                             \
@@ -597,13 +598,16 @@
         /* We earlier created 'sv' for very large version numbers, to rely  \
          * on the specialized algorithms SV code has built-in for such      \
          * values */                                                        \
+        double d = SvNVX(ver); \
         if (sv) {                                                           \
-            Perl_sv_setpvf(aTHX_ sv, "%.9" NVff, SvNVX(ver));               \
+            Perl_sv_setpvf(aTHX_ sv, "%d.%09d", (int) d,                    \
+                (int) ((long long)(d * 1000000000) % 1000000000));          \
             len = SvCUR(sv);                                                \
             buf = SvPVX(sv);                                                \
         }                                                                   \
         else {                                                              \
-            len = my_snprintf(tbuf, sizeof(tbuf), "%.9" NVff, SvNVX(ver));  \
+            len = my_snprintf(tbuf, sizeof(tbuf), "%d.%09d", (int) d,       \
+                (int) ((long long)(d * 1000000000) % 1000000000));          \
             buf = tbuf;                                                     \
         }                                                                   \
                                                                             \
