diff -Paur --no-dereference -- bochs.upstream/Makefile.in bochs/Makefile.in
--- bochs.upstream/Makefile.in
+++ bochs/Makefile.in
@@ -35,7 +35,7 @@
 top_builddir    = .
 top_srcdir      = $(srcdir)
 
-DESTDIR =
+DESTDIR ?=
 
 VERSION=@VERSION@
 REL_STRING=@REL_STRING@
@@ -655,6 +655,8 @@
 	@CD_UP_THREE@
 	@RMCOMMAND@ Makefile
 
+distclean: dist-clean
+
 ###########################################
 # Build app on MacOS X
 ###########################################
diff -Paur --no-dereference -- bochs.upstream/configure bochs/configure
--- bochs.upstream/configure
+++ bochs/configure
@@ -25275,7 +25275,7 @@
   if test "$with_sdl" = yes; then
     SOUNDLOW_OBJS="$SOUNDLOW_OBJS soundsdl.o"
     if test "$bx_plugins" = 1; then
-      SDL_SOUND_LINK_OPTS="$SDL_SOUND_LINK_OPTS `sdl-config --libs`"
+      SDL_SOUND_LINK_OPTS="$SDL_SOUND_LINK_OPTS `${PKG_CONFIG:-pkg-config} sdl --libs`"
     fi
     soundlow_drivers="$soundlow_drivers sdl"
     if test "$soundlow_default" = "dummy"; then
@@ -25931,14 +25931,14 @@
   SPECIFIC_GUI_OBJS="$SPECIFIC_GUI_OBJS \$(GUI_OBJS_SDL)"
   # GUI_*FLAGS are added to the compilation of every bochs file, not just
   # the files in gui/*.cc.
-  SDL_CFLAGS=`sdl-config --cflags`
+  SDL_CFLAGS=`${PKG_CONFIG:-pkg-config} sdl --cflags`
   GUI_CFLAGS="$GUI_CFLAGS $SDL_CFLAGS"
   GUI_CXXFLAGS="$GUI_CXXFLAGS $SDL_CFLAGS"
-  GUI_LINK_OPTS_SDL="`sdl-config --libs`"
+  GUI_LINK_OPTS_SDL="`${PKG_CONFIG:-pkg-config} sdl --libs`"
   GUI_LINK_OPTS="$GUI_LINK_OPTS \$(GUI_LINK_OPTS_SDL)"
   # The plugin version uses multi-threading support in Bochs core
   if test "$bx_plugins" = 1; then
-    NONPLUGIN_GUI_LINK_OPTS="`sdl-config --libs`"
+    NONPLUGIN_GUI_LINK_OPTS="`${PKG_CONFIG:-pkg-config} sdl --libs`"
   fi
   # The enhanced X debugger depends on GTK2
   if test "$gui_debugger" = 1 -a "$DEFAULT_GUI" != win32; then
@@ -27196,8 +27196,8 @@
 main (void)
 {
 pthread_t th; pthread_join(th, 0);
-		     pthread_attr_init(0); pthread_cleanup_push(0, 0);
-		     pthread_create(0,0,0,0); pthread_cleanup_pop(0);
+		     pthread_attr_init(0);
+		     pthread_create(0,0,0,0);
   ;
   return 0;
 }
diff -Paur --no-dereference -- bochs.upstream/doc/docbook/Makefile.in bochs/doc/docbook/Makefile.in
--- bochs.upstream/doc/docbook/Makefile.in
+++ bochs/doc/docbook/Makefile.in
@@ -26,7 +26,7 @@
 DOCBOOK2HTML = $(JADE) $(JADE_ARGS)
 endif
 
-DESTDIR=
+DESTDIR?=
 
 # name of the major documentation sections
 SECTIONS=user documentation development
diff -Paur --no-dereference -- bochs.upstream/gui/textconfig.cc bochs/gui/textconfig.cc
--- bochs.upstream/gui/textconfig.cc
+++ bochs/gui/textconfig.cc
@@ -48,6 +48,7 @@
 #include <ctype.h>
 #include <string.h>
 #include <assert.h>
+#include <inttypes.h>
 
 #ifndef __QNXNTO__
 }
@@ -205,7 +206,7 @@
       bx_printf("Your choice must be an integer between " FMT_LL "d and " FMT_LL "d.\n\n", min, max);
       continue;
     }
-    illegal = (1 != sscanf(buffer, "%ld", &n));
+    illegal = (1 != sscanf(buffer, "%" PRId64, &n));
     if (illegal || n<min || n>max) {
       bx_printf("Your choice (%s) was not an integer between " FMT_LL "d and " FMT_LL "d.\n\n",
              clean, min, max);
diff -Paur --no-dereference -- bochs.upstream/iodev/hdimage/vbox.cc bochs/iodev/hdimage/vbox.cc
--- bochs.upstream/iodev/hdimage/vbox.cc
+++ bochs/iodev/hdimage/vbox.cc
@@ -138,7 +138,7 @@
   // (10gig = 10240 1-meg blocks with each entry using 4 bytes) = 40k
   mtlb = new Bit32s[(unsigned) header.blocks_in_hdd];
   if (mtlb == 0) {
-    BX_PANIC(("unable to allocate %lu bytes for vbox image's map table", header.blocks_in_hdd * sizeof(Bit32u)));
+    BX_PANIC(("unable to allocate %zu bytes for vbox image's map table", (size_t) header.blocks_in_hdd * sizeof(Bit32u)));
   }
 
   // read in the map table
diff -Paur --no-dereference -- bochs.upstream/iodev/hpet.cc bochs/iodev/hpet.cc
--- bochs.upstream/iodev/hpet.cc
+++ bochs/iodev/hpet.cc
@@ -391,7 +391,7 @@
   }
   if (diff < HPET_MIN_ALLOWED_PERIOD) diff = HPET_MIN_ALLOWED_PERIOD;
   if (diff > HPET_MAX_ALLOWED_PERIOD) diff = HPET_MAX_ALLOWED_PERIOD;
-  BX_DEBUG(("Timer %d to fire in 0x%lX ticks", t->tn, diff));
+  BX_DEBUG(("Timer %d to fire in 0x%lX ticks", t->tn, (long) diff));
   bx_pc_system.activate_timer_nsec(t->timer_id, ticks_to_ns(diff), 0);
 }
 
diff -Paur --no-dereference -- bochs.upstream/iodev/serial.cc bochs/iodev/serial.cc
--- bochs.upstream/iodev/serial.cc
+++ bochs/iodev/serial.cc
@@ -30,6 +30,18 @@
 // is used to know when we are exporting symbols and when we are importing.
 #define BX_PLUGGABLE
 
+#if defined(__sortix__)
+struct hostent
+{
+	char   *h_name;
+	char  **h_aliases;
+	int     h_addrtype;
+	int     h_length;
+	char  **h_addr_list;
+};
+#define gethostbyname(a) ((void) a, (struct hostent*) NULL)
+#endif
+
 #include "iodev.h"
 
 #if !defined(WIN32) || defined(__CYGWIN__)
@@ -489,9 +501,9 @@
 
         memset ((char*) &sin, 0, sizeof (sin));
 #if BX_HAVE_SOCKADDR_IN_SIN_LEN
-       sin.sin_len = sizeof sin;
+        sin.sin_len = sizeof sin;
 #endif
-        memcpy ((char*) &(sin.sin_addr), hp->h_addr, hp->h_length);
+        memcpy ((char*) &(sin.sin_addr), hp->h_addr_list[0], hp->h_length);
         sin.sin_family = hp->h_addrtype;
         sin.sin_port = htons (port);
 
