diff -Paur --no-dereference -- vim.upstream/src/Makefile vim/src/Makefile
--- vim.upstream/src/Makefile
+++ vim/src/Makefile
@@ -2305,6 +2305,8 @@
 # install targets
 
 install: $(GUI_INSTALL)
+	ln -s ../../etc/vimrc $(DEST_VIM)/vimrc
+	ln -s ../../etc/gvimrc $(DEST_VIM)/gvimrc
 
 install_normal: installvim installtools $(INSTALL_LANGS) install-icons
 
diff -Paur --no-dereference -- vim.upstream/src/sha256.c vim/src/sha256.c
--- vim.upstream/src/sha256.c
+++ vim/src/sha256.c
@@ -407,10 +407,7 @@
     char_u	     sha256sum[32];
     context_sha256_T ctx;
 
-    srand(get_some_time());
-
-    for (i = 0; i < (int)sizeof(random_data) - 1; i++)
-	random_data[i] = (char_u)((get_some_time() ^ rand()) & 0xff);
+    arc4random_buf(random_data, sizeof(random_data));
     sha256_start(&ctx);
     sha256_update(&ctx, (char_u *)random_data, sizeof(random_data));
     sha256_finish(&ctx, sha256sum);
